runtime/compiler/optimizer/J9RecognizedCallTransformer.hpp

```diff
@@ -139,6 +139,82 @@ class RecognizedCallTransformer : public OMR::RecognizedCallTransformer
     *     \endcode
     */
    void process_java_lang_StringUTF16_toBytes(TR::TreeTop* treetop, TR::Node* node);
+   /** \brief
+    *     Transforms jdk/internal/util/ArraysSupport.vectorizedMismatch(Ljava/lang/Object;JLjava/lang/Object;JII)I
+    *     into an arraycmp, bit manipulation and iselect sequence with equivalent semantics.
+    *
+    *  \param treetop
+    *     The treetop which anchors the call node.
+    *
+    *  \param node
+    *     The call node representing a call to jdk/internal/util/ArraysSupport.vectorizedMismatch(Ljava/lang/Object;JLjava/lang/Object;JII)I
+    *     which has the following shape:
+    *
+    *     \code
+    *     icall  <jdk/internal/util/ArraysSupport.vectorizedMismatch(Ljava/lang/Object;JLjava/lang/Object;JII)I>
+    *       <a>
+    *       <aOffset>
+    *       <b>
+    *       <bOffset>
+    *       <length>
+    *       <log2ArrayIndexScale>
+    *     \endcode
+    *
+    *  \details
+    *     The call node is transformed to the following shape:
+    *
+    *     \code
+    *     iselect ()
+    *       icmpeq
+    *         arraycmp (arrayCmpLen )
+    *           aladd
+    *             <a>
+    *             <aOffset>
+    *           aladd
+    *             <b>
+    *             <bOffset>
+    *           i2l
+    *             iand
+    *               ishl
+    *                 <length>
+    *                 <log2ArrayIndexScale>
+    *               ixor
+    *                 ior
+    *                   ishl
+    *                     <log2ArrayIndexScale>
+    *                     iconst 1
+    *                   iconst 3
+    *                 iconst -1
+    *         ==>iand
+    *       ixor
+    *         ishr
+    *           iand
+    *             ==>ishl
+    *             ==>ior
+    *           <log2ArrayIndexScale>
+    *         iconst -1
+    *       ishr
+    *         ==>arraycmp
+    *         <log2ArrayIndexScale>
+    *     \endcode
+    *
+    *     This transformation is valid because vectorizedMismatch is functionally equivalent to the following pseudocode,
+    *     assuming log2ArrayIndexScale <= 3.
+    *
+    *     \code
+    *     vectorizedMismatch(a, aOffset, b, bOffset, length, log2ArrayIndexScale) {
+    *         lengthInBytes = length << log2ArrayIndexScale
+    *         mask = (log2ArrayIndexScale<<1) | 3
```

Comment:
I didn't have a problem with the IL generated for the mask. Expressing it with shift/or was fine, and it probably generates better code than `iselect`. The unclear part (IMO) was why the mask has that value. So I was just looking for a comment in the code that says something along the lines of "yes, even though this method can compare any multiple of 4 bytes, it will restrict itself to multiples of 8 bytes when the elements are 32-bit"

Reply:
This is not to say that you necessarily need to put the IL back the way it was, by the way. It looks like it's equivalent, and I expect that it barely affects the benefit of this transformation.

Reply:
Oh, I see you've added a comment as well, thanks! Ideally I'd prefer for it to clarify that this is in fact the behaviour of the original method, e.g. if you decide you do want to put the the IL back how it was anyway.

But if you'd rather not change it anymore, I think I'm OK accepting the comment as-is because it at least makes it clear that this is the behaviour that you intended.

Reply:
I added a bit more explanation in the comment while I was updating the generated tree